AWSCognitoIdentityProvider/AWSCognitoIdentityUserPool.m

```diff
@@ -61,12 +62,26 @@ + (instancetype)defaultCognitoIdentityUserPool {
     dispatch_once(&onceToken, ^{
         AWSServiceConfiguration *serviceConfiguration = nil;
         AWSServiceInfo *serviceInfo = [[AWSInfo defaultAWSInfo] defaultServiceInfo:AWSInfoCognitoUserPool];
-
         if (serviceInfo) {
-            serviceConfiguration = [[AWSServiceConfiguration alloc] initWithRegion:serviceInfo.region
-                                                               credentialsProvider:nil];
+            NSString *endpointOverride = [serviceInfo.infoDictionary objectForKey:AWSCognitoUserPoolEndpoint];
+            if (endpointOverride) {
+                NSURLComponents *components = [NSURLComponents new];
+                components.scheme = @"https";
+                components.host = endpointOverride;
+                NSURL *endpointURL = [components URL];
```

Comment:
Should we check if endpointURL is nil and throw exception?

Reply:
Good callout--that gives us a fail-fast path during config instead of waiting until the first call is made somewhere into the runtime.

Reply:
Note revision. `NSURLComponents` would actually create a valid URL from a `host` string like `https://foo.com`--it would percent-escape the string and the resulting URL would be `https://https%2A%3F%3Ffoo.com`, which of course fails to resolve at runtime. The revised code assigns the incoming `Endpoint` value to `percentEscapedHost`, which will throw an exception if it receives an unescaped string. That, coupled with the validation check on the existence of the URL, should give us a reasonable balance between fail-fast detection and being overly clever on our validation and missing a valid host case. I have not updated the CHANGELOG instructions to note that the hostname must be escaped, since valid hostnames are only supposed to contain latin alphanumerics, hyphens, and dot-separators, none of which are escaped.